src/tree.rs

```diff
@@ -102,42 +135,116 @@ fn cond_swap(swap: bool, node: pallas::Base, sibling: pallas::Base) -> Pair {
 /// The layer with 2^n nodes is called "layer n":
 ///      - leaves are at layer MERKLE_DEPTH_ORCHARD = 32;
 ///      - the root is at layer 0.
-/// `l_star` is MERKLE_DEPTH_ORCHARD - layer - 1.
+/// `l` is MERKLE_DEPTH_ORCHARD - layer - 1.
 ///      - when hashing two leaves, we produce a node on the layer above the leaves, i.e.
-///        layer = 31, l_star = 0
-///      - when hashing to the final root, we produce the anchor with layer = 0, l_star = 31.
-fn hash_layer(l_star: usize, pair: Pair) -> pallas::Base {
+///        layer = 31, l = 0
+///      - when hashing to the final root, we produce the anchor with layer = 0, l = 31.
+fn hash_with_l(l: usize, pair: Pair) -> CtOption<pallas::Base> {
     // MerkleCRH Sinsemilla hash domain.
     let domain = HashDomain::new(MERKLE_CRH_PERSONALIZATION);
 
-    domain
-        .hash(
-            iter::empty()
-                .chain(i2lebsp_k(l_star).iter().copied())
-                .chain(
-                    pair.left
-                        .to_le_bits()
-                        .iter()
-                        .by_val()
-                        .take(L_ORCHARD_MERKLE),
-                )
-                .chain(
-                    pair.right
-                        .to_le_bits()
-                        .iter()
-                        .by_val()
-                        .take(L_ORCHARD_MERKLE),
-                ),
-        )
-        .unwrap()
+    domain.hash(
+        iter::empty()
+            .chain(i2lebsp_k(l).iter().copied())
+            .chain(
+                pair.left
+                    .to_le_bits()
+                    .iter()
+                    .by_val()
+                    .take(L_ORCHARD_MERKLE),
+            )
+            .chain(
+                pair.right
+                    .to_le_bits()
+                    .iter()
+                    .by_val()
+                    .take(L_ORCHARD_MERKLE),
+            ),
+    )
+}
+
+/// A newtype wrapper for leaves and internal nodes in the Orchard
+/// incremental note commitment tree.
+#[derive(Clone, Debug)]
+pub struct OrchardIncrementalTreeDigest(CtOption<pallas::Base>);
+
+impl OrchardIncrementalTreeDigest {
+    /// Creates an incremental tree life digest for the specified
+    /// Pallas base field element.
+    pub fn from_cmx(value: &ExtractedNoteCommitment) -> Self {
+        OrchardIncrementalTreeDigest(CtOption::new(**value, 1.into()))
+    }
+
+    /// Convert this digest to its canonical byte representation.
+    pub fn to_bytes(&self) -> Option<[u8; 32]> {
+        <Option<pallas::Base>>::from(self.0).map(|b| b.to_bytes())
+    }
+
+    /// Parses a incremental tree leaf digest from the bytes of
+    /// a note commitment.
+    pub fn from_bytes(bytes: &[u8; 32]) -> Self {
```

Comment:
It's an error in the caller to pass an invalid (not canonical Fp) `bytes` value, so that should be distinguished from the hash returning ⊥.